lib/tokenize.cpp

```diff
@@ -136,6 +136,14 @@ const Token * Tokenizer::isFunctionHead(const Token *tok, const std::string &end
             tok = tok->next();
         if (Token::Match(tok, "= 0|default|delete ;"))
             tok = tok->tokAt(2);
+        if (Token::simpleMatch(tok, "requires")) {
+            for (tok = tok->next(); tok && !Token::Match(tok, ";|{"); tok = tok->next()) {
```

Comment:
I don't like the loop condition `tok && !Token::Match(tok, ";|{")` would it be possible to invert it somehow to some `Token::Match(tok, "..expected..")`? I don't know how this keyword works.. can it be more or less arbitrary expressions that the loop should skip?

Reply:
> I don't like the loop condition tok && !Token::Match(tok, ";|{") would it be possible to invert it somehow to some Token::Match(tok, "..expected..")? 

I could try and see if `Token::Match(tok, "!![;|{]")` works, but usually these are not composable.

> I don't know how this keyword works.. can it be more or less arbitrary expressions that the loop should skip?

This is finding the end of the `requires` statement which stops at a end of statement(ie `;`) or the beginning of the function body(ie `{`).  This is similar to the loop used above for the trailing return types.

Reply:
I meant that "..expected.." should be something like "%name%|.." where we try to explicitly say what types of tokens we skip.

just a wild idea if there is something like this :

     MACRO(foo() requires true);

it would be preferable if the isFunctionHead returns false even though there is ; after the requires.

Reply:
> I meant that "..expected.." should be something like "%name%|.." where we try to explicitly say what types of tokens we skip.

This doesn't seem very robust either, if this allows more tokens than we listed in the future or we miss listing token because we dont have a test case for it.

> 
> just a wild idea if there is something like this :
> 
> ```
>  MACRO(foo() requires true);
> ```
> 
> it would be preferable if the isFunctionHead returns false even though there is ; after the requires.

We could stop at closing braces as well, I am not sure if `isFuncionHead` would return false in this case. Either way the loop for the trailing return type would need to be updated as well.

Reply:
I updated the matcher to stop also at closing braces.